### PR TITLE
Added support for the ThreadNamesList minidump stream

### PR DESCRIPTION
**Pull request by gabrielesvelto**
This fixes issue #44

### CODE REVIEW

**Comment by willkg** [comment]
I spent some time looking into the situation. Right now, thread names in the stackwalk output are determined by mapping the thread names from the `ThreadIdNameMapping` crash annotation to the thread index. That works fine for Windows and Linux. [Bug 1658831](https://bugzilla.mozilla.org/show_bug.cgi?id=1658831) covers how this doesn't work on OSX.

Meanwhile, we discovered that there's a `ThreadNamesStream` in the minidump on Windows. So we fixed the Linux and OSX code so now they populate that stream as well in minidumps.

This PR adds some code to use the `ThreadNamesStream` to determine the thread names for the stackwalk output instead of using the `ThreadIdNameMapping`. That moves us another step to fixing it on all three platforms. It's also nice because it removes one of the uses of the crash annotations data. If we remove all uses, then we don't need to pass that file in anymore. (This may remove the one and only use. I forget where that's at.)

After we land this, I'll look into whether anything needs to change in Socorro in addition to picking up a new release of minidump-stackwalk.

To test this, I want to pull down minidumps for a recent Windows, Linux, and OSX crash, run them through this minidump-stackwalk passing in crash annotations that don't include the `ThreadIdNameMapping` and verify we get thread names for all three platforms.

**Comment by willkg** [comment]
I tested with crash reports from each platform for a recent Firefox nightly. I did two runs:

1. ran `./bin/run_mdsw.sh` passing in the `--raw-json` argument (this is the crash annotations)
2. ran `./bin/run_mdsw.sh` without passing in a `--raw-json` argument

* windows: e8a6ff49-db45-4f29-8b3c-357ee0210715
  * printed thread names with and without the crash annotation data
* osx: 83aacb88-1d6f-4078-8e19-ecd3e0210715
  * printed thread names with and without the crash annotation data
* linux: 786e2964-c727-43b9-b097-5e70c0210715 (https://crash-stats.mozilla.org/report/index/786e2964-c727-43b9-b097-5e70c0210715#tab-details)
  * prints thread names when passed the crash annotations data
  * does not print thread names without crash annotations data

That last one puzzles me. Shouldn't minidumps from a recent Firefox nightly on Linux be populating the `ThreadNamesList` stream? Is that a bug in the fix for [bug #1714465](https://bugzilla.mozilla.org/show_bug.cgi?id=1714465) or a bug in this code (or did I pick a terrible example crash report to test with)?

**Comment by gabrielesvelto** [comment]
> That last one puzzles me. Shouldn't minidumps from a recent Firefox nightly on Linux be populating the `ThreadNamesList` stream? Is that a bug in the fix for [bug #1714465](https://bugzilla.mozilla.org/show_bug.cgi?id=1714465) or a bug in this code (or did I pick a terrible example crash report to test with)?

@gabrielesvelto uncloaks from PTO like a Romulan warbird

I forgot that on Linux it's only supported in content process crashes for now, not main process crashes.

@gabrielesvelto cloaks back into PTO

**Comment by willkg** [comment]
Sorry--I didn't realize we weren't trying to remove support for extra.json anymore (https://github.com/luser/rust-minidump/issues/177). I'll stop testing along those lines.

File: minidump-stackwalk/stackwalker.cc [hunk 1]
```diff
@@ -917,7 +917,11 @@ static void ConvertProcessStateToJSON(const ProcessState& process_state,
       }
       thread["last_error_value"] = last_error_value_str;
     }
-    auto thread_name = thread_id_name_map[raw_stack->tid()];
+    auto thread_name = raw_stack->name();
+    if (thread_name.empty()) {
+      thread_name = thread_id_name_map[raw_stack->tid()];
+    }
```
Comment (willkg) [inline minidump-stackwalk/stackwalker.cc:923]: This is going to prefer the value in the `ThreadNamesList` stream over the one provided in the raw crash extra json file. Figured I'd call that out so future me can find it faster if need be.

**Comment by willkg** [comment]
I went through a whole bunch of crash reports looking at whether this causes segfaults with minidump-stackwalk and what kinds of crash reports it affects.

I didn't see any segfaults--mdsw worked fine.

Here are some differences:
* Bunch of Windows crashes now have thread names:
  * 16b5888c-7156-4f83-bc20-d67fe0210719 (78.12.0 ESR)
  * 1ca71ac2-bc44-463f-b496-4a79c0210719 (78.12.0 ESR)
  * 035b304a-6589-4333-9abd-c25d10210719 (90.0 release)
  * 73bc0c80-3067-4b23-9a00-962c40210719 (90.0 release
* One of the Linux crash reports changed thread names:
  * b5d5497b-7aed-4835-bde7-9bdb80210719 "TaskController Thr" to "TaskCon-read #0"
* Some MacOS crash reports have thread names now:
  * 220574d6-b75c-48bc-8ee7-19f7f0210720 (91.0 aurora 20210713163746)
  * 40ba6029-2fd5-4283-8f26-688790210720 (92.0a1 nightly 20210715094037)
  * 5574cd2c-3acd-48f6-9540-0137e0210720 (91.0 beta 20210713163746)
  * 76f9de1a-5470-49af-91ea-73e2e0210720 (92.0a1 nightly 20210715094037)

Generally, the situation looks better with the one exception being that one Linux crash report where the thread names changed. They're still readable and I'm pretty sure that's because the source of the threadnames is different. So I think that's fine.

The code changes look fine as far as I could tell.

I think this is good to land and we can let it run on Socorro stage for a while and see how it fares.

**Review by willkg — APPROVED** [review]

**Pull request merged**